src/modules/roc_core/target_linux/roc_core/timer.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2020 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <errno.h>
+#include <unistd.h>
+
+#include "roc_core/errno_to_str.h"
+#include "roc_core/panic.h"
+#include "roc_core/timer.h"
+
+namespace roc {
+namespace core {
+
+#if defined(CLOCK_MONOTONIC)
+#define CLOCK_TYPE CLOCK_MONOTONIC
+#else
+#define CLOCK_TYPE CLOCK_REALTIME
+#endif
+
+Timer::Timer()
+    : timerfd_(timerfd_create(CLOCK_TYPE, 0))
+    , deadline_(0) {
+    if (timerfd_ == -1) {
+        roc_panic("timer: timerfd_create(CLOCK_TYPE, 0): %s", errno_to_str().c_str());
+    }
+}
+
+Timer::~Timer() {
+    close(timerfd_);
+}
+
+bool Timer::try_set_deadline(nanoseconds_t new_deadline) {
+    if (!deadline_.try_store(new_deadline)) {
+        return false;
+    }
+
+    struct itimerspec new_value = convert_deadline();
+
+    int res = timerfd_settime(timerfd_, TFD_TIMER_ABSTIME, &new_value, NULL);
+    if (res == -1) {
+        roc_panic("timer: timerfd_settime(timerfd_, TFD_TIMER_ABSTIME, "
+                  "&new_value, NULL): %s",
+                  errno_to_str().c_str());
+    }
+
+    return true;
+}
+
+void Timer::wait_deadline() {
+    const nanoseconds_t deadline = deadline_.wait_load();
+
+    if (deadline >= 0 && deadline <= timestamp()) {
+        return;
+    }
+
+    uint64_t ticks = 0;
+    ssize_t readed = read(timerfd_, &ticks, sizeof(uint64_t));
+    if (readed == -1 || ticks == 0) {
+        roc_panic("timer: read(timerfd_, &ticks, sizeof(uint64_t)): %s",
+                  errno_to_str().c_str());
+    }
+}
+
+struct itimerspec Timer::convert_deadline() {
+    const nanoseconds_t deadline = deadline_.wait_load();
```

Comment:
Since we call wait_load() here, try_set_deadline() is not wait-free any more. We should get rid of this call.

Reply:
ok